### ISSUE TITLE
ability to delete built-in zones

### ISSUE THREAD

**Issue by laywah2016** [t-woerner/firewalld#139]
Currently there are multiple built-in zones which can't be deleted.

root@fedora]# firewall-cmd --delete-zone=work --permanent
Error: BUILTIN_ZONE: 'work' is built-in zone

FedoraServer
FedoraWorkstation
block
dmz
drop
external
home
internal (active)
public
trusted
work

as this creates a an overly sophisticated and annoying iptables list when trying to display iptables it would be a good idea to be able to delete those that are unused.

**Comment by t-woerner** [comment]
Zones are only created in netfilter if they are used. Used here means that there is at least a source or interface bound to that zone of if it is the default zone.
Which firewalld version are you using?

**Comment by laywah2016** [comment]
Installed Packages
Name        : firewalld
Arch        : noarch
Epoch       : 0
Version     : 0.4.0
Release     : 2.fc23

**Comment by t-woerner** [comment]
Please add the output of `firewall-cmd --get-active-zones` to identify why the zones are created in netfilter. Please also add the output of `iptables-save | grep -- "-A INPUT_ZONES"`.

**Comment by laywah2016** [comment]
Sorry for the delay in responding.
This is the output you asked for.

# firewall-cmd --get-active-zones
internal
  interfaces: ens3


# iptables-save | grep -- "-A INPUT_ZONES"
-A INPUT_ZONES -i ens3 -g IN_internal
-A INPUT_ZONES -g IN_internal

**Comment by t-woerner** [comment]
This means there is only the internal zone created in the firewall rules, the others are not in the iptables-save or iptables -L output.
Do you have another reason for building zone removal?

**Comment by laywah2016** [comment]
I understand that they may not be saved using iptables-save. I would like to know however if they can be removed completely if they are not needed.
For me, I feel it just complicates the view when doing a iptables -L and you have a bunch of zones showing that are not needed.

**Comment by t-woerner** [comment]
Ununsed zones are not visible with iptables, because there are no rules generated for them with newer versions.
firewalld is not removing zones from the /usr/lib path. The zones in /usr/lib are part of the package.

**Comment by t-woerner** [comment]
firewalld is not allowing to remove the built-in zones to make sure that the user is not removing required zones like for example the public zone. It would be possible to remove the dmz, external, home, internal and work zones without loosing the ability to start firewalld. But for example the block, drop and trusted zones are the base zones. These are the minimal requirements that are checked by firewalld at start. The public zone (default zone setting from firewalld.conf) is used if it is available, otherwise the external and then the block zone will become the default zone.

**Comment by olebowle** [comment]
>  Do you have another reason for building zone removal?

To get detailed information about the zone we're using, I usually issue `firewall-cmd --list-all-zones`. Unfortunately, also inactive zones are shown this way. I could use `firewall-cmd --get-active-zones`, but this output is not detailed enough. That's why it would be nice to be able to delete shipped zones.

As a crude workaround I use something like this: `firewall-cmd --list-all-zones | awk '!/^[[:blank:]]/ && /active/ {p=1} !/^[[:blank:]]/ && !/active/ {p=0} p'` showing me detailed information about only the active zones.

tl;dr
I guess what I'm really asking for is `firewall-cmd --get-active-zones --list-all`, which doesn't work.

**Comment by ppokorny** [comment]
Yes, please.  Add to 4.4.4 or later release (For Red Hat and sibling distro) either:  --get-active-zones --list-all or let us delete "built-in" rules that aren't in use (work, home, public, dmz) so that it's easier to audit what the rules actually implement.

Being able to restore "built-in" rules would also be needed.

Perhaps treat zones like services.  With some pre-defined ones that are available, but only if you --add-zone.

**Comment by fessmage** [comment]
I vote also for option like `--list-all-zones` but which outputs active zones only. Something like `--list-all-active-zones`.
